### ISSUE TITLE
(plasma_spin.h) plasma_spin_pause(): Add aarch64 support and use isb instead of yeild for it

### ISSUE THREAD

**Issue by a-p-jo** [gstrauss/plasma#1]
Firstly, the main motivation for this issue: `isb` does better on aarch64 what `yield` does on arm.
See:
- https://web.archive.org/web/20231004132033/https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8258604
- https://bugs.mysql.com/bug.php?id=100664

Now, onto the code suggestions:

1. With msvc, you only detect x86, amd64 and ia64. Now that windows on arm is getting serious (though it has been a thing for long), it might be worthwhile to throw in something like:

```c
// detect old 32-bit arm and 16-bit arm-thumb
#elif defined (_M_ARM) || defined(_M_ARMT)
// https://learn.microsoft.com/en-us/cpp/intrinsics/arm-intrinsics
		#define plasma_spin_pause() __yield()
#elif defined(_M_ARM64) || defined(_M_ARM64EC)
// https://learn.microsoft.com/en-us/cpp/intrinsics/arm64-intrinsics
		#define plasma_spin_pause() __isb()
```
  arm64ec is an ABI for windows 11 arm apps that allows interop with amd64 code running in emulation, see: https://learn.microsoft.com/en-us/windows/arm/arm64ec

2. With gnu c, you handle arm with just an `#elif defined(__arm__)`, and emitting a `yield`.
Firstly, `yield` is available in thumb mode too, so you may want to instead do `#elif defined(__arm__) || defined(__thumb__)`.

Secondly, iirc `__arm__` does not identify aarch64. So consider throwing in a:
```c
#elif defined(__aarch64___)
		#define plasma_spin_pause() __asm__("isb")
```
Lastly, `__volatile__` is unnecessary here, because in gnu c "asm statements that have no output operands... are implicitly volatile", see: https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#Volatile.

**Comment by gstrauss** [comment]
Thank you for all those details and references!  I'll have some time to look into this next week.

(BTW, I prefer to have `__volatile__` explicit unless doing so is somehow harmful.)

**Comment by a-p-jo** [comment]
@gstrauss had the time to look into it yet? I find your library interesting for work on a thread-safe freestanding allocator, and might use it if it can support arm and aarch64 pauses in its spinloops

**Comment by gstrauss** [comment]
ARM architecture has advanced quite a bit since I wrote this over a decade ago.  At that time, `__aarch64__` was new, and I did not have any hardware on which to test.  Similarly, I do not have `__thumb__` hardware.  This library does not currently contain code specifically for `__aarch64__` or `__thumb__`.

Thank you for the notes about yield and isb on aarch64.

To support these architectures properly, I need to modify `plasma_spin.h` as well as `plasma_membar.h`.  However, since the introduction of C11/C++11 atomics, the need to use assembly for memory barriers is reduced, so I'll incorporate your suggestions for `plasma_spin_pause()` for now, and leave the rest for another time.

How might I thank you in the commit message?  Name, email, or simply github tag a-p-jo?
